src/Core/Services/MetadataProviders/SqlMetadataProvider.cs

```diff
@@ -1830,73 +1885,99 @@ IEnumerable<List<ForeignKeyDefinition>> foreignKeysForAllTargetEntities
             List<ForeignKeyDefinition> validatedFKDefinitionsToTarget = new();
-            foreach (ForeignKeyDefinition fKDefinitionToTarget in fKDefinitionsToTarget)
+            foreach (ForeignKeyDefinition configResolvedFkDefinition in fKDefinitionsToTarget)
             {
-                // This code block adds FK definitions between source and target entities when there is an FK constraint defined
-                // in the database, either from source->target or target->source entities or both.
-
-                // Add the referencing and referenced columns for this foreign key definition for the target.
+                // Check whether DAB found a match between 'configResolvedFkDefinition' and 'databaseResolvedFKDefinition's {referencing -> referenced} entity pair.
                 if (PairToFkDefinition is not null &&
-                    PairToFkDefinition.TryGetValue(fKDefinitionToTarget.Pair, out ForeignKeyDefinition? inferredFKDefinition))
+                    PairToFkDefinition.TryGetValue(configResolvedFkDefinition.Pair, out ForeignKeyDefinition? databaseResolvedFkDefinition))
                 {
-                    // Being here indicates that we inferred an FK constraint for the current foreign key definition.
-                    // The count of referencing and referenced columns being > 0 indicates that source.fields and target.fields 
-                    // have been specified in the config file.
-                    // In this scenario, higher precedence is given to the fields configured through the config file. So, the existing FK definition is retained as is.
-                    if (fKDefinitionToTarget.ReferencingColumns.Count > 0 && fKDefinitionToTarget.ReferencedColumns.Count > 0)
+                    if (DoesConfiguredRelationshipOverrideDatabaseFkConstraint(configResolvedFkDefinition))
                     {
-                        validatedFKDefinitionsToTarget.Add(fKDefinitionToTarget);
+                        validatedFKDefinitionsToTarget.Add(configResolvedFkDefinition);
+
+                        // Save additional metadata for use when processing requests on self-joined/referencing entities.
+                        if (IsSelfJoiningRelationship(configResolvedFkDefinition))
+                        {
+                            EntityRelationshipKey entityToFkDefKey = new(
+                                entityName: configResolvedFkDefinition.SourceEntityName,
+                                relationshipName: configResolvedFkDefinition.RelationshipName);
+                            RelationshipToFkDefinition.TryAdd(entityToFkDefKey, configResolvedFkDefinition);
+                        }
                     }
-                    // The count of referenced and referencing columns being = 0 indicates that source.fields and target.fields 
-                    // are not configured through the config file. In this case, the FK fields inferred from the database are populated.
                     else
```

Comment:
why dont we add the `configResolvedFkDefinition` to validatedFKdefinitions in this case since that has more metadata?

Reply:
Good catch, both the configResolvedFkDefinition and databaseResolvedFkDefinition had important metadata that would be beneficial to consolidation.

![image](https://github.com/Azure/data-api-builder/assets/6414189/9ec003ab-1b68-454e-a61d-e71ea7d04f43)

I added this change to consolidate the referencing/referenced columns from `databaseResolvedFkDefinition` into `configResolvedFkDefinition` because `configResolvedFkDefinition` was missing that metadata, but did have the metadata ( the "more metadata" you called out) for relationshipName/sourceEntityName.

```csharp
                        // When the configured relationship doesn't override the database FK constraint,
                        // DAB can consolidate the referenced and referencing columns from the database FK definition
                        // into the configResolvedFkDefinition object.
                        configResolvedFkDefinition.ReferencedColumns = databaseResolvedFkDefinition.ReferencedColumns;
                        configResolvedFkDefinition.ReferencingColumns = databaseResolvedFkDefinition.ReferencingColumns;
                        validatedFKDefinitionsToTarget.Add(configResolvedFkDefinition);

                        // Save additional metadata for use when processing requests on self-joined/referencing entities.
                        // Since the configResolvedFkDefinition has additional metadata populated, DAB supplements that
                        // object by using the inferred FK definition's referenced/referencing columns.
                        if (IsSelfJoiningRelationship(databaseResolvedFkDefinition))
                        {
                            EntityRelationshipKey entityToFkDefKey = new(
                                entityName: configResolvedFkDefinition.SourceEntityName,
                                relationshipName: configResolvedFkDefinition.RelationshipName);
                            RelationshipToFkDefinition.TryAdd(entityToFkDefKey, configResolvedFkDefinition);
                        }
```